Moves features tab to release coordinator page

lkbhitesh07 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of N/A.
2. This PR does the following: As part of the [project: Normalize the usage of platform-parameters](https://docs.google.com/document/d/1ndjDml4PUB_LA8mwwTQOawDFOWyNnhOiPXpAWaK2cXI/edit) this PR moves the features tab to the release-coordinator page.

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

**The features tab is on the release-coordinator page**
![Screenshot from 2023-06-02 12-17-56](https://github.com/oppia/oppia/assets/44300735/20e07196-c6cb-4551-8bc4-19aa7174c6b7)

**Video that shows the features tab works as expected**
Steps I followed in the video:
1. On the learner-dashboard page I tried to visit the "blog" page and it redirected me to the medium article.
2. I then visited the features tab on the release coordinator and then turned on the "blog" feature to be visible on the dev server.
3. Then I reloaded the learner-dashboard page and again tried visiting the "blog" page and I was able to access it.
4. I then visited the admin page to show how it looks after updates.

https://github.com/oppia/oppia/assets/44300735/54999eb5-1f50-4653-89e2-2f37c8b2693c

**Loading screen when the feature flags are getting fetched from the backend**

https://github.com/oppia/oppia/assets/44300735/dafa91c5-4483-4ad1-8902-b2efd8cb0963

**Loading screen when switching between tabs**

https://github.com/oppia/oppia/assets/44300735/98a3d9f1-a559-44fd-acd1-035687acb7b3


## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

lkbhitesh07 committed:
initial changes

lkbhitesh07 committed:
added the feature handler at backend and made the changes in the frontend

lkbhitesh07 committed:
fixed lint

lkbhitesh07 committed:
intial changes for e2e test file for release-coordinator

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into move-features-tab-to-release-coordinator-page

lkbhitesh07 committed:
niy

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into move-features-tab-to-release-coordinator-page

lkbhitesh07 committed:
introduced webdriverio file for release coordinator and shifted the code from AdminPage js

lkbhitesh07 committed:
fixed lint

lkbhitesh07 committed:
nit change in e2e test

lkbhitesh07 committed:
nits to fix e2e tests

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into move-features-tab-to-release-coordinator-page

lkbhitesh07 committed:
nit to fix the e2e test

lkbhitesh07 committed:
corrected the hover-over-name of features tab

lkbhitesh07 commented:
Hey @seanlip, can you please take the first pass?
Thanks

seanlip commented on core/templates/domain/platform_feature/platform-feature-admin-backend-api.service.ts:
```diff
@@ -35,11 +45,27 @@ export class PlatformFeatureAdminBackendApiService {
       Promise<void> {
     await this.http.post(
-      AdminPageConstants.ADMIN_HANDLER_URL,
+      '/feature_flags',
```
Should use a named constant here and elsewhere.

lkbhitesh07 replied:
Done

seanlip commented on main.py:
```diff
@@ -601,6 +601,9 @@ def get_redirect_route(
     get_redirect_route(
         r'/memorycachehandler', release_coordinator.MemoryCacheHandler),
 
```
Maybe drop this newline to clump the release_coordinator stuff together.

lkbhitesh07 replied:
Done

seanlip reviewed (changes requested):
Thanks @lkbhitesh07! Not many comments, this looks generally fine. Three notes about your video though:

- I don't see how, midway through the video, you changed the blog page to point back to localhost?
- Can you show the /admin page as well?
- When you first click to the "Features" tab, it says there are no features, but this is wrong. Try to prevent that, e.g. say "Loading feature flags..." or, perhaps better, use the "Loading ..." interstitial that we have elsewhere in the codebase. But don't show wrong information here.

Thanks!

seanlip reviewed (changes requested):
Thanks @lkbhitesh07! Not many comments, this looks generally fine. Three notes about your video though:

- I don't see how, midway through the video, you changed the blog page to point back to localhost?
- Can you show the /admin page as well?
- When you first click to the "Features" tab, it says there are no features, but this is wrong. Try to prevent that, e.g. say "Loading feature flags..." or, perhaps better, use the "Loading ..." interstitial that we have elsewhere in the codebase. But don't show wrong information here.

Thanks!

lkbhitesh07 committed:
created constant for the url and tried adding loading screen

lkbhitesh07 committed:
nits

lkbhitesh07 commented:
Hey @seanlip, thanks for the review.



> I don't see how, midway through the video, you changed the blog page to point back to localhost?

I did not realize that only part of the screen was getting recorded. I have now updated the video and also mentioned the steps I perform for the testing. It will be clear now.
I was actually switching tabs in between and that was not recorded in the video.


> Can you show the /admin page as well?

I have shown this in the video.


> When you first click to the "Features" tab, it says there are no features, but this is wrong. Try to prevent that, e.g. say "Loading feature flags..." or, perhaps better, use the "Loading ..." interstitial that we have elsewhere in the codebase. But don't show wrong information here.

I'm a little bit stuck at this part. I tried implementing it but the "loading" screen does not seem to appear. Changes I made:

[features-tab.component.html](https://github.com/oppia/oppia/blob/9e5d1e681590e7d9d9fc70ca9df1402642c97e93/core/templates/pages/release-coordinator-page/features-tab/features-tab.component.html#L195-L197)
Added a `showLoadingScreen` flag so that the below div only appears when there are no feature flags.
```
<div *ngIf="featureFlags.length === 0 && !showLoadingScreen" class="oppia-page-card oppia-long-text oppia-admin-features oppia-empty-indicator e2e-test-no-feature">
  No platform features are available.
</div>
```

[features-tab.component.ts](https://github.com/oppia/oppia/blob/9e5d1e681590e7d9d9fc70ca9df1402642c97e93/core/templates/pages/release-coordinator-page/features-tab/features-tab.component.ts#L140-L149)
I tried adding the loading screen here but I'm not able to see it. Please note that the delays in the loading of the features-tab is not consistent it only happens sometimes. Below are the changes I made to the function that fetches the feature-flags:

```
async reloadFeatureFlagsAsync(): Promise<void> {
    this.loaderService.showLoadingScreen('Loading');
    this.showLoadingScreen = true;
    const data = await this.apiService.getFeatureFlags();
    this.featureFlags = data.featureFlags;
    this.featureFlagNameToBackupMap = new Map(
      this.featureFlags.map(feature => [[feature.name](http://feature.name/), cloneDeep(feature)]));
    this.showLoadingScreen = false;
    this.loaderService.hideLoadingScreen();
  }
```
I tried looking at the implementation for this in the codebase and I was following a similar procedure but I'm not able to get the loading screen. Could you please provide some guidance?

Thanks

**Edit:** I also tried using `then` to fetch the feature-flags, I refactored the code for the function as:
```
    this.loaderService.showLoadingScreen('Loading');
    this.showLoadingScreen = true;
    this.apiService.getFeatureFlags().then((data) => {
      this.featureFlags = data.featureFlags;
      this.loaderService.hideLoadingScreen();
      this.showLoadingScreen = false;
    });
```

lkbhitesh07 committed:
fix for e2e test

lkbhitesh07 commented:
@U8NWXD @gp201 @Rijuta-s @JayVivarekar @aasiffaizal PTAL

lkbhitesh07 committed:
nit fix for the e2e test

seanlip commented:
Hi @lkbhitesh07, thanks. Video looks good. I am not sure your loading screen logic is correct though. I would do the following:

- Change the var name to loadingScreenIsShown -- booleans should not start with verbs. Keeping that convention helps you differentiate clearly between booleans and methods, and avoid confusing yourself -- note that your code also uses showLoadingScreen() as a function.
- When the page/tab initializes, set loadingScreenIsShown to true
- Once the feature flags have loaded from the backend, set loadingScreenIsShown to false

See if that helps!

lkbhitesh07 commented:
Thanks @seanlip, I tried making changes locally but it did not seem to work and I'm not sure why. So the `ngOnInit` function looks something like the below:
We wanted to start the loading screen when the component gets initialized.
```
ngOnInit(): void {
    this.loadingScreenIsShown = true;
    this.loaderService.showLoadingScreen('Loading');
    this.reloadFeatureFlagsAsync();
    this.reloadDummyHandlerStatusAsync();
  }
```

The `reloadFeatureFlagsAsync` function looks something like:
```
async reloadFeatureFlagsAsync(): Promise<void> {
    const data = await this.apiService.getFeatureFlags();
    this.loadingScreenIsShown = false;
    this.featureFlags = data.featureFlags;
    this.featureFlagNameToBackupMap = new Map(
      this.featureFlags.map(feature => [feature.name, cloneDeep(feature)]));
    this.loaderService.hideLoadingScreen();
  }
```
Here I'm calling the `hideLoadingScreen` function. I have also tried shifting positions for `hideLoadingScreen` part here and there.

The complete component file for the changes can be found [here](https://github.com/oppia/oppia/blob/c908306ab83a9723414701c9b8dd204a9c01cf85/core/templates/pages/release-coordinator-page/features-tab/features-tab.component.ts), please note that the latest changes are not present which are changes for the `ngOnInit` and `reloadFeatureFlagsAsync` as I have not pushed the changes and I'm still trying to figure it out.

Just to confirm we do not have to make any changes to the `html` file in order to display the loading screen, right? I have taken look at other files and I have not located any changes to the html file regarding this.
I'm sorry if I missed out on something, I'm still getting familiarize with the frontend concepts.

Thanks

lkbhitesh07 committed:
tried adding loading screen

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into move-features-tab-to-release-coordinator-page

lkbhitesh07 commented:
Hey guys(@U8NWXD @gp201 @Rijuta-s @JayVivarekar @aasiffaizal) we plan to wrap up this PR soon as other PRs will be dependent on this one. Also, I and Sean are having a discussion on the loading screen part, feel free to provide some guidance.
Thanks in advance.

seanlip commented:
@lkbhitesh07 I'm pretty sure you'll need to make changes to the html template. If you look at base-content.component.html (which is the parent component for most of the Oppia pages) you'll see there's an oppia-loading-message; look at the logic for it. Then try to look through what changes happen in the JS/TS code when loadingMessage is set to '' when loading completes.

However I think the admin/release-coordinator pages don't derive from base-content.component.html (not sure?) so the logic there might need to be different.

seanlip commented:
Basically for your approach you might need some kind of `<ng-if="loadingScreenIsShown">` clause in the HTML.

U8NWXD reviewed (approved):
Note that I am reviewing on behalf of @oppia/automated-qa-reviewers and @oppia/dev-workflow-reviewers since @seanlip has already reviewed for @oppia/release-workflow-reviewers

lkbhitesh07 committed:
added the loading screen on the features tab

lkbhitesh07 commented:
Thanks, @U8NWXD for the quick response.

Thanks @seanlip for the suggestion, I was able to locate the logic for the loading screen part. I have made the changes and now the loading screen seems to appear, I have added the video in the description.

@seanlip PTAL

seanlip commented on core/controllers/release_coordinator.py:
```diff
@@ -49,3 +54,121 @@ def get(self) -> None:
         self.render_json({})
+
+
+class FeatureFlagsNormalizePayloadDict(TypedDict):
```
Normalize --> Normalized?

Also the name here doesn't make sense, a normalized payload dict should correspond to a handler, not "feature flags". Feature flags don't have an action, for example.

lkbhitesh07 replied:
Updated

seanlip commented on core/controllers/release_coordinator.py:
```diff
@@ -49,3 +54,121 @@ def get(self) -> None:
+    dictionary.
+    """
+
+    action: Optional[str]
```
Why are these optional? Shouldn't they be required?

lkbhitesh07 replied:
I have updated this.

seanlip commented on core/controllers/release_coordinator.py:
```diff
@@ -49,3 +54,121 @@ def get(self) -> None:
+                        'update_feature_flag_rules'
+                    ]
+                },
+                # TODO(#13331): Remove default_value when it is confirmed that,
```
Not sure what this means in this context.

lkbhitesh07 replied:
Yeah, this does not makes sense here. Removed

seanlip commented on core/controllers/release_coordinator.py:
```diff
@@ -49,3 +54,121 @@ def get(self) -> None:
+                    feature_services.FeatureFlagNotFoundException) as e:
+                raise self.InvalidInputException(e)
+
+            new_rule_dicts = [rules.to_dict() for rules in new_rules]
```
rule.to_dict() for rule in new_rules

(unless each item in the list is multiple rules?)

lkbhitesh07 replied:
Updated

seanlip commented on core/templates/pages/release-coordinator-page/features-tab/features-tab.component.ts:
```diff
@@ -129,25 +127,26 @@ export class AdminFeaturesTabComponent implements OnInit {
-    const data = await this.adminDataService.getDataAsync();
-
+    const data = await this.apiService.getFeatureFlags();
+    this.loadingScreenIsShown = false;
```
Why do you need both this and hideLoadingScreen() below?

lkbhitesh07 replied:
The `loadingScreenIsShown` variable is for the following part:
```
<div *ngIf="featureFlags.length === 0 && !loadingScreenIsShown" class="oppia-page-card oppia-long-text oppia-admin-features oppia-empty-indicator e2e-test-no-feature">
  No platform features are available.
</div>
```
We would require to do this in our HTML part so that the `No platform features are available` part only occurs when there are no platform params. If we remove `loadingScreenIsShown` check, then instead of the loading screen `No platform features are available` part appears. I tried it out locally.

seanlip replied:
But semantically, "this.loadingScreenIsShown = false" and "this.loaderService.hideLoadingScreen()" sound like they are doing similar things. So I'm asking why we need to call both of them. What does each contribute? I think you justified the first but not the second.

If they are doing different things, can you find a way to differentiate the names so that it's clear what role each plays?

Also by "check" in your previous comment, do you mean "assignment"? It doesn't seem like a check to me, just a direct variable assignment.

lkbhitesh07 replied:
I see. I have renamed the variable to `featureFlagsAreFetched`.
So, `loaderService.showLoadingScreen('Loading')` and `loaderService.hideLoadingScreen()` is used to show and hide the loading screen. The `featureFlagsAreFetched` variable is used to keep track of whether the feature flags are fetched completely from the backend or not, So now the `No platform features are available` part will only appear when the `featureFlagsAreFetched` variable is set to `true` and the `featureFlags.length === 0`.

> Also by "check" in your previous comment, do you mean "assignment"? It doesn't seem like a check to me, just a direct variable assignment.

By check, I was referring to the `*ngIf` we have on the HTML page. I was basically referring to the removal of the variable from the `*ngIf`. Sorry for not being clear.

seanlip reviewed (changes requested):
Thanks @lkbhitesh07 -- UI looks fine. One check: could you please switch to the feature flags tab, switch away, and then switch back to it? Would like to see the second-time loading experience (there probably shouldn't be any if the data's already in the frontend, but if we do indeed do another fetch then there should be another load screen).

Also left some comments inline, PTAL.

seanlip reviewed (changes requested):
Thanks @lkbhitesh07 -- UI looks fine. One check: could you please switch to the feature flags tab, switch away, and then switch back to it? Would like to see the second-time loading experience (there probably shouldn't be any if the data's already in the frontend, but if we do indeed do another fetch then there should be another load screen).

Also left some comments inline for codeowner files, PTAL.

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into move-features-tab-to-release-coordinator-page

lkbhitesh07 committed:
done suggested changes

lkbhitesh07 committed:
fixed frontend tests

lkbhitesh07 committed:
fixed lint

lkbhitesh07 commented:
> Thanks @lkbhitesh07 -- UI looks fine. One check: could you please switch to the feature flags tab, switch away, and then switch back to it? Would like to see the second-time loading experience (there probably shouldn't be any if the data's already in the frontend, but if we do indeed do another fetch then there should be another load screen).
> 
> Also left some comments inline for codeowner files, PTAL.

I have updated the PR description with the video. @seanlip PTAL

seanlip commented on core/controllers/release_coordinator.py:
```diff
@@ -163,10 +161,10 @@ def post(self) -> None:
                 raise self.InvalidInputException(e)
 
-            new_rule_dicts = [rules.to_dict() for rules in new_rules]
+            new_rules_dicts = [rule.to_dict() for rule in new_rules]
```
new_rule_dicts is fine, the idea is that "rule_dict" --> "rule_dicts" when pluralized. Otherwise it sounds like each dict has multiple rules.

lkbhitesh07 replied:
Got it, I have reverted this.

lkbhitesh07 committed:
renamed the loadingScreenIsShown variable

lkbhitesh07 commented:
@seanlip PTAL

lkbhitesh07 commented:
Hey @JayVivarekar @gp201 @Rijuta-s @aasiffaizal, Just a friendly ping in case you missed my comment.
Thanks

seanlip reviewed (approved):
LGTM. Thanks!

JayVivarekar reviewed (approved):
LGTM

Rijuta-s reviewed (approved):
Looks good to me for code owner files!

gp201 reviewed (approved):
LGTM

lkbhitesh07 commented:
Hey @aasiffaizal, Just a ping in case you missed my earlier comments.
Thanks

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into move-features-tab-to-release-coordinator-page

aasiffaizal reviewed (approved):
LGTM
